File: sdk/java/src/main/java/jp/kengos/shojiku/ShojikuException.java
package jp.kengos.shojiku;

/**
 * The base of every exception this package throws.
 *
 * <p>Throwing is deliberately rare here. A template that will not render, a key that will not sign,
 * a signature that does not verify are OUTCOMES — they come back as {@link Result} objects you
 * query, never as exceptions you catch. What is left for exceptions is what every Java library
 * reserves them for: programmer misuse, and an environment that cannot host the engine at all.
 *
 * <p>Unchecked, deliberately: none of these is a condition a correct program recovers from at the
 * call site, and a checked exception would push {@code try}/{@code catch} into code whose whole
 * point is that failure arrives as data.
 */
public class ShojikuException extends RuntimeException {

  /**
   * Creates the exception.
   *
   * @param message what went wrong
   */
  public ShojikuException(String message) {
    super(message);
  }

  /**
   * Creates the exception over the one underneath it.
   *
   * @param message what went wrong
   * @param cause the exception this one wraps
   */
  public ShojikuException(String message, Throwable cause) {
    super(message, cause);
  }
}

File: sdk/php/src/Exception/ShojikuException.php
<?php

declare(strict_types=1);

namespace Shojiku\Exception;

/**
 * The marker every exception this package throws implements.
 *
 * An INTERFACE rather than a base class, because PHP already has the right
 * bases and they carry meaning: `\LogicException` is what the language
 * reserves for a mistake in the calling program, `\RuntimeException` for a
 * condition only the environment can produce. Inheriting from one shared
 * base would throw that distinction away, so each exception here extends the
 * SPL class its meaning belongs to and implements this instead — a caller
 * that wants "anything Shojiku threw" catches this, and a caller that wants
 * "a bug in my own code" catches `\LogicException`.
 *
 * Throwing is deliberately rare. A template that will not render, a key that
 * will not sign, a signature that does not verify are OUTCOMES — they come
 * back as {@see \Shojiku\Result} objects you query, never as exceptions you
 * catch.
 */
interface ShojikuException extends \Throwable
{
}
